CAS Unittests broken on Windows

```
FAILED: unittests/CAS/CMakeFiles/CASTests.dir/CASTestConfig.cpp.obj
C:\PROGRA~2\MICROS~2\2022\BUILDT~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\cl.exe  /nologo /TP -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_LIBCPP_ENABLE_HARDENED_MODE -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -IS:\b\1\unittests\CAS -IS:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS -IS:\b\1\include -IS:\SourceCache\swift-project\llvm-project\llvm\include -IS:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include -IS:\SourceCache\swift-project\llvm-project\third-party\unittest\googlemock\include /GS- /Gw /Gy /Oi /Oy /Zc:inline /Zi /Zc:__cplusplus /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Zi /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -std:c++17 -MD  /EHs-c- /GR- -UNDEBUG /showIncludes /Founittests\CAS\CMakeFiles\CASTests.dir\CASTestConfig.cpp.obj /Fdunittests\CAS\CMakeFiles\CASTests.dir\ /FS -c S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\CASTestConfig.cpp
S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\CASTestConfig.cpp(30): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\CASTestConfig.cpp(30): error C2065: 'constructor': undeclared identifier
S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\CASTestConfig.cpp(30): error C2062: type 'void' unexpected                                                                                                                                                                 S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\CASTestConfig.cpp(30): error C2143: syntax error: missing ';' before '{'                                                                                                                                                   S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\CASTestConfig.cpp(30): error C2447: '{': missing function header (old-style formal list?)                                                                                                                                  [315/864] Building CXX object unittests\CAS\CMakeFiles\CASTests.dir\ObjectStoreTest.cpp.obj                                                                                                                                                                                             S:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include\gtest/internal/custom/raw-ostream.h(68): warning C4180: qualifier applied to function type has no meaning; ignored                                                                                    S:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include\gtest/internal/custom/raw-ostream.h(34): note: see reference to class template instantiation 'llvm_gtest::StreamSwitch<T,void>' being compiled
        with
        [
            T=void (llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,uint64_t)
        ]
S:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include\gtest/gtest-message.h(132): note: see reference to function template instantiation 'decltype(auto) llvm_gtest::printable<T>(T (__cdecl &))' being compiled
        with
        [
            T=void (llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,uint64_t)
        ]
S:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include\gtest/gtest.h(2223): note: see reference to function template instantiation 'testing::Message &testing::Message::operator <<<T>(T (__cdecl &))' being compiled
        with
        [
            T=void (llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,uint64_t)
        ]
S:\SourceCache\swift-project\llvm-project\llvm\unittests\CAS\ObjectStoreTest.cpp(286): note: see reference to function template instantiation 'testing::ScopedTrace::ScopedTrace<void(llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,uint64_t)>(const char *,int,T (__cdecl &))' being compiled
        with
        [
            T=void (llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,uint64_t)
        ]
S:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include\gtest/internal/custom/raw-ostream.h(53): warning C4180: qualifier applied to function type has no meaning; ignored
S:\SourceCache\swift-project\llvm-project\third-party\unittest\googletest\include\gtest/internal/custom/raw-ostream.h(36): note: see reference to class template instantiation 'llvm_gtest::RawStreamProxy<T>' being compiled
        with
        [
            T=void (llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,llvm::cas::ObjectStore &,uint64_t)
        ]
[323/864] Building CXX object unittests\CodeGen\CMakeFiles\CodeGenTests.dir\AArch64SelectionDAGTest.cpp.obj
```

--- a/llvm/lib/CAS/OnDiskCommon.cpp
+++ b/llvm/lib/CAS/OnDiskCommon.cpp
@@ -9,18 +9,40 @@
 #include "OnDiskCommon.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/Support/Error.h"
+#include "llvm/Support/Process.h"
+#include <mutex>
+#include <optional>
 
 using namespace llvm;
 
+static uint64_t OnDiskCASMaxMappingSize = 0;
+
 Expected<std::optional<uint64_t>> cas::ondisk::getOverriddenMaxMappingSize() {
-  constexpr const char *EnvVar = "LLVM_CAS_MAX_MAPPING_SIZE";
-  const char *Value = getenv(EnvVar);
-  if (!Value)
+  static std::once_flag Flag;
+  Error Err = Error::success();
+  std::call_once(Flag, [&Err] {
+    ErrorAsOutParameter EAO(&Err);
+    constexpr const char *EnvVar = "LLVM_CAS_MAX_MAPPING_SIZE";
+    auto Value = sys::Process::GetEnv(EnvVar);
+    if (!Value)
+      return;
+
+    uint64_t Size;
+    if (StringRef(*Value).getAsInteger(/*auto*/ 0, Size))
+      Err = createStringError(inconvertibleErrorCode(),
+                              "invalid value for %s: expected integer", EnvVar);
+    OnDiskCASMaxMappingSize = Size;
+  });
+
+  if (Err)
+    return std::move(Err);
+
+  if (OnDiskCASMaxMappingSize == 0)
     return std::nullopt;
 
-  uint64_t Size;
-  if (StringRef(Value).getAsInteger(/*auto*/ 0, Size))
-    return createStringError(inconvertibleErrorCode(),
-                             "invalid value for %s: expected integer", EnvVar);
-  return Size;
+  return OnDiskCASMaxMappingSize;
+}
+
+void cas::ondisk::setMaxMappingSize(uint64_t Size) {
+  OnDiskCASMaxMappingSize = Size;
 }

--- a/llvm/lib/CAS/OnDiskCommon.h
+++ b/llvm/lib/CAS/OnDiskCommon.h
@@ -14,10 +14,16 @@
 
 namespace llvm::cas::ondisk {
 
-/// Retrieves an overridden maximum mapping size for CAS files, if any, by
-/// checking LLVM_CAS_MAX_MAPPING_SIZE in the environment. If the value is
-/// unreadable, returns an error.
+/// Retrieves an overridden maximum mapping size for CAS files, if any,
+/// speicified by LLVM_CAS_MAX_MAPPING_SIZE in the environment or set by
+/// `setMaxMappingSize()`. If the value from environment is unreadable, returns
+/// an error.
 Expected<std::optional<uint64_t>> getOverriddenMaxMappingSize();
+
+/// Set MaxMappingSize for ondisk CAS. This function is not thread-safe and
+/// should be set before creaing any ondisk CAS and does not affect CAS already
+/// created. Set value 0 to use default size.
+void setMaxMappingSize(uint64_t Size);
 
 } // namespace llvm::cas::ondisk
 

--- a/llvm/unittests/CAS/CASTestConfig.cpp
+++ b/llvm/unittests/CAS/CASTestConfig.cpp
@@ -10,6 +10,7 @@
 #include "llvm/CAS/ActionCache.h"
 #include "llvm/CAS/ObjectStore.h"
 #include "llvm/RemoteCachingService/RemoteCachingService.h"
+#include "llvm/Support/CommandLine.h"
 #include "gtest/gtest.h"
 #include <mutex>
 
@@ -27,23 +28,20 @@ TestingAndDir createInMemory(int I) {
   return TestingAndDir{std::move(CAS), std::move(Cache), nullptr, std::nullopt};
 }
 
-__attribute__((constructor)) static void configureCASTestEnv() {
-  // Restrict the size of the on-disk CAS for tests. This allows testing in
-  // constrained environments (e.g. small TMPDIR). It also prevents leaving
-  // behind large files on file systems that do not support sparse files if a
-  // test  crashes before resizing the file.
-  static std::once_flag Flag;
-  std::call_once(Flag, [] {
-    size_t Limit = 100 * 1024 * 1024;
-    std::string LimitStr = std::to_string(Limit);
-    setenv("LLVM_CAS_MAX_MAPPING_SIZE", LimitStr.c_str(), /*overwrite=*/false);
-  });
-}
-
 INSTANTIATE_TEST_SUITE_P(InMemoryCAS, CASTest,
                          ::testing::Values(createInMemory));
 
 #if LLVM_ENABLE_ONDISK_CAS
+namespace llvm::cas::ondisk {
+void setMaxMappingSize(uint64_t Size);
+} // namespace llvm::cas::ondisk
+
+void setMaxOnDiskCASMappingSize() {
+  static std::once_flag Flag;
+  std::call_once(
+      Flag, [] { llvm::cas::ondisk::setMaxMappingSize(100 * 1024 * 1024); });
+}
+
 TestingAndDir createOnDisk(int I) {
   unittest::TempDir Temp("on-disk-cas", /*Unique=*/true);
   std::unique_ptr<ObjectStore> CAS;
@@ -55,6 +53,8 @@ TestingAndDir createOnDisk(int I) {
                        std::move(Temp)};
 }
 INSTANTIATE_TEST_SUITE_P(OnDiskCAS, CASTest, ::testing::Values(createOnDisk));
+#else
+void setMaxOnDiskCASMappingSize() {}
 #endif /* LLVM_ENABLE_ONDISK_CAS */
 
 #if LLVM_CAS_ENABLE_REMOTE_CACHE

--- a/llvm/unittests/CAS/CASTestConfig.h
+++ b/llvm/unittests/CAS/CASTestConfig.h
@@ -33,6 +33,8 @@ struct TestingAndDir {
   std::optional<llvm::unittest::TempDir> Temp;
 };
 
+void setMaxOnDiskCASMappingSize();
+
 class CASTest
     : public testing::TestWithParam<std::function<TestingAndDir(int)>> {
 protected:
@@ -58,7 +60,10 @@ protected:
       Envs.emplace_back(std::move(TD.Env));
     return std::move(TD.Cache);
   }
-  void SetUp() { NextCASIndex = 0; }
+  void SetUp() {
+    NextCASIndex = 0;
+    setMaxOnDiskCASMappingSize();
+  }
   void TearDown() {
     NextCASIndex = std::nullopt;
     Dirs.clear();

--- a/llvm/unittests/CAS/ObjectStoreTest.cpp
+++ b/llvm/unittests/CAS/ObjectStoreTest.cpp
@@ -416,6 +416,7 @@ TEST(OnDiskCASTest, BlobsBigParallelMultiCAS) {
 
 #ifndef _WIN32 // FIXME: resize support on Windows.
 TEST(OnDiskCASTest, DiskSize) {
+  setMaxOnDiskCASMappingSize();
   unittest::TempDir Temp("on-disk-cas", /*Unique=*/true);
   std::unique_ptr<ObjectStore> CAS;
   ASSERT_THAT_ERROR(createOnDiskCAS(Temp.path()).moveInto(CAS), Succeeded());

